client/verta/verta/dataset/_s3.py

```diff
@@ -93,22 +111,66 @@ def _get_s3_metadata(cls, bucket_name, key=None):
         s3 = boto3.client('s3')
 
         # TODO: handle prefixes
-        if key is None:
+        if s3_loc.key is None:
             # TODO: handle `bucket_name` not found
-            for obj in s3.list_objects(Bucket=bucket_name)['Contents']:
-                yield cls._get_s3_obj_metadata(obj, bucket_name, obj['Key'])
+            for obj in s3.list_object_versions(Bucket=s3_loc.bucket)['Versions']:
+                if obj['IsLatest']:
+                    yield cls._get_s3_obj_metadata(obj, s3_loc.bucket, obj['Key'])
         else:
             # TODO: handle `key` not found
-            obj = s3.head_object(Bucket=bucket_name, Key=key)
-            yield cls._get_s3_obj_metadata(obj, bucket_name, key)
+            if s3_loc.version_id is not None:
+                # TODO: handle `version_id` not found
+                obj = s3.head_object(Bucket=s3_loc.bucket, Key=s3_loc.key, VersionId=s3_loc.version_id)
+            else:
+                obj = s3.head_object(Bucket=s3_loc.bucket, Key=s3_loc.key)
+            yield cls._get_s3_obj_metadata(obj, s3_loc.bucket, s3_loc.key)
 
     @classmethod
     def _get_s3_obj_metadata(cls, obj, bucket_name, key):
         # pylint: disable=no-member
         msg = _DatasetService.S3DatasetComponentBlob()
         msg.path.path = cls._S3_PATH.format(bucket_name, key)
-        msg.path.size = obj.get('Size') or obj['ContentLength']
+        msg.path.size = obj.get('Size') or obj.get('ContentLength') or 0
         msg.path.last_modified_at_source = _utils.timestamp_to_ms(_utils.ensure_timestamp(obj['LastModified']))
         msg.path.md5 = obj['ETag'].strip('"')
+        if obj.get('VersionId', 'null') != 'null':
+            msg.s3_version_id = obj['VersionId']
 
         return msg
+
+    @staticmethod
+    def location(bucket, key=None, version_id=None):
```

Comment:
I would recommend parsing S3 url here (or in the constructor) too. I don't see a reason to 1) force different interfaces, 2) force the user to split.